state-sync/src/coordinator.rs

```diff
@@ -811,92 +869,97 @@ impl<T: ExecutorProxyTrait> StateSyncCoordinator<T> {
-    /// return Ok(()) if above checks all pass and chunk was stored to storage
-    fn apply_chunk(&mut self, peer: &PeerNetworkId, response: GetChunkResponse) -> Result<()> {
+    /// Applies (i.e., executes and stores) the chunk to storage iff `response` is valid.
+    fn apply_chunk(
+        &mut self,
+        peer: &PeerNetworkId,
+        response: GetChunkResponse,
+    ) -> Result<(), Error> {
         debug!(
             LogSchema::event_log(LogEntry::ProcessChunkResponse, LogEvent::Received)
                 .chunk_response(response.clone())
                 .peer(peer)
         );
         fail_point!("state_sync::apply_chunk", |_| {
-            Err(anyhow::anyhow!("Injected error in apply_chunk"))
+            Err(crate::error::Error::UnexpectedError(
+                "Injected error in apply_chunk".into(),
+            ))
         });
 
+        // Check response comes from upstream peer
         if !self.request_manager.is_known_upstream_peer(peer) {
             counters::RESPONSE_FROM_DOWNSTREAM_COUNT
                 .with_label_values(&[
                     &peer.raw_network_id().to_string(),
                     &peer.peer_id().to_string(),
                 ])
                 .inc();
-            bail!("received chunk response from downstream");
+            return Err(Error::ReceivedChunkFromDownstream(peer.to_string()));
         }
 
+        // Check chunk is not empty
         let txn_list_with_proof = response.txn_list_with_proof.clone();
-        let known_version = self.local_state.synced_version();
         let chunk_start_version =
             txn_list_with_proof
                 .first_transaction_version
                 .ok_or_else(|| {
                     self.request_manager.process_empty_chunk(&peer);
-                    format_err!("[state sync] Empty chunk from {:?}", peer)
+                    Error::ReceivedEmptyChunk(peer.to_string())
                 })?;
 
+        // Check chunk starts at the correct version
+        let known_version = self.local_state.synced_version();
         let expected_version = known_version
             .checked_add(1)
-            .ok_or_else(|| format_err!("Expected version has overflown!"))?;
+            .ok_or_else(|| Error::IntegerOverflow("Expected version has overflown!".into()))?;
         if chunk_start_version != expected_version {
-            // Old / wrong chunk.
             self.request_manager.process_chunk_version_mismatch(
                 peer,
                 chunk_start_version,
                 known_version,
             )?;
         }
 
-        let chunk_size = txn_list_with_proof.len() as u64;
+        // Process the chunk based on the response type
         match response.response_li {
             ResponseLedgerInfo::VerifiableLedgerInfo(li) => {
-                self.process_response_with_verifiable_li(txn_list_with_proof, li, None)
+                self.process_response_with_verifiable_li(txn_list_with_proof.clone(), li, None)
```

Comment:
Why do we clone the object, just so we can get the length at the end rather than the beginning?

It doesn't change the chunksize from before to after because you clone it, and has some memory / performance implications.

Reply:
Personally, I think this is debatable:
- My preference is to group code together (as much as reasonably possible), hence why I dislike creating variables early on in a function and not using them for 50+ lines 😄 Granted, though, I think all of this code needs to be cleaned up. These methods are long and like spaghetti.
- Regarding performance, I don't think this is significant and I'd argue it's a premature optimization right now. What we might lose by doing 1-2 clones per chunk response, we gain in clarity.

Overall, this code won't live like this for long, and right now I'm in favor of choosing clarity over a (small?) performance hit.